### PR TITLE
[Backend] Alerting Part 2 & Refactoring

### LINKED ISSUE #380: Username/Password for zabbix credentials available for everyone who can see dashboard
- [ x] Bug report

#### For bug report please include this information:
- What Grafana version are you using? 4.2.0
- What Zabbix version are you using? 2.4
- What zabbix plugin version are you using? 3.3.0
- What OS are you running grafana on? Debian 4.6-6 (2016-07-26)
- What did you do?
When I open any zabbix dashboard I see ajax requests with zabbix user & password inside:
https://host/api/datasources/proxy/23
`{jsonrpc: "2.0", method: "user.login", params: {user: "XXX", password: "XXXXXXXXX"}, id: 1}`
- What was the expected result?
User & password allready stored in datasource. There is no reason show it to any user viewing dashboard.
- What happened instead?
Every user can see zabbix user/password if has access to view dashboard

### PR DESCRIPTION
**Pull request by alecxvs**
This PR fully connects the functions for alerting on the backend. It's not feature complete yet but it can perform basic alerting queries.

There's a large refactoring as part of this. We divided the code into three files: backend, datasource, and api_client. We actually did this some time ago, so the creation of zabbix_api_core sort of follows this same logic. If that name is better, it shouldn't be difficult to rename api_client to zabbix_api_core.

The purpose for refactoring was mostly to make unit testing more concise. It's not possible to mock methods within the same struct, so when any one test runs, it depends on a lot of various state. We determined a good way to split the functions is by plugin api, datasource logic, and zabbix api. For instance, it should be easier now to test that the filtering logic in getItems is correct because we can mock the API Client to check the exact ids it's calling with, and we don't need to build HTTP mocks in every test.

Basic outline of the refactored files:
- Backend contains ZabbixPlugin, and not much has changed here besides some fixes. 
- Datasource contains ZabbixDatasource and has the main query entrypoint functions (ZabbixAPIQuery, TimeseriesQuery -- which alerting uses and in the future all metrics queries will use, and TestConnection) as well as the rest of the individual queries that contain filtering logic but don't directly build requests to the Zabbix API
- API Client contains all of the methods that are used to call Zabbix. These form the queries based on given IDs and parameters, and then on return, marshal the data into a known type

Other changes:
- Introducing more types to the frontend
- Migrate frontend to secure json. Finally closes #380. This means no longer carrying username/password, and the datasource configuration will save to secure json. The credential hiding/reset stuff was mostly copied from the CloudWatch datasource in Grafana.

This PR should contain all the tests from #844 (albeit in a different form). I'll confirm that with @vignesh-reddy first though.

### CODE REVIEW

**Comment by vignesh-reddy** [comment]
Yes, this refactoring PR contains all the test from #844

File: pkg/datasource.go [hunk 1]
```diff
@@ -5,123 +5,478 @@ import (
...
+	dsInfo *datasource.DatasourceInfo
+	hash   string
+	logger hclog.Logger
 }
 
-func (p *ZabbixPlugin) NewZabbixDatasource(dsInfo *datasource.DatasourceInfo) (*ZabbixDatasource, error) {
-	ds, err := newZabbixDatasource(dsInfo)
+// NewZabbixDatasource returns an instance of ZabbixDatasource with an API Client
+func NewZabbixDatasource(logger hclog.Logger, dsInfo *datasource.DatasourceInfo) (*ZabbixDatasource, error) {
+	client, err := NewZabbixAPIClient(logger, dsInfo.GetUrl())
```
Comment (alexanderzobnin) [inline pkg/datasource.go]: Something goes wrong here - ZabbixAPIClient required datasource set for correct work, but ZabbixDatasource requred ZabbixAPIClient. Looks like circular dependency.

**Comment by alexanderzobnin** [comment]
Update: `canary` version required until 6.6 release at least. I've found a reason of bug in tests and opened an issue
https://github.com/grafana/grafana/issues/21537

File: pkg/backend.go [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+			&datasource.QueryResult{
+				RefId:    "zabbixAPI",
+				MetaJson: string(jsonBytes),
+			},
+		},
+	}, nil
+}
+
+// BuildMetricsResponse builds a response object using a given TimeSeries array
+func BuildMetricsResponse(results []*datasource.QueryResult) (*datasource.DatasourceResponse, error) {
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply pkg/backend.go]: af14296

File: pkg/datasource.go (hunk 1 above)
Comment (alecxvs) [reply pkg/datasource.go]: Yeah, it's definitely a circular reference. It was a compromise to keep the changes you made that remove dsInfo from all of the function parameters. A great improvement but not something that we resolved in our refactor.

Need to double check what ZabbixAPIClient needs from ZabbixDatasource. If it's just dsInfo, maybe we can store that on the client instead.

File: src/datasource-zabbix/types.ts [hunk 3]
```diff
@@ -13,7 +13,43 @@ export interface ZabbixConnectionTestQuery {
   queryType: string;
 }
 
+export interface ZabbixJsonData extends DataSourceJsonData {
```
[the remark this answers is not in the document]
Comment (alecxvs) [reply src/datasource-zabbix/types.ts]: 962b4f6

**Comment by alecxvs** [comment]
@alexanderzobnin I'm trying to test changes on the frontend in regards to the username field, but I'm getting a weird error that's not showing up in the logs or console. It seems Grafana or React might be hiding it. Have you seen this before?

```
Error: Super expression must either be null or a function Evaluating https://.../public/plugins/alexanderzobnin-zabbix-app/datasource-zabbix/module.js Loading plugins/alexanderzobnin-zabbix-app/datasource-zabbix/module
```

**Comment by alexanderzobnin** [comment]
@alecxvs could you post a difference between code that throws error and valid?

**Comment by alecxvs** [comment]
I'm still trying to find a valid version. I just pulled the current backend branch and tried deploying that but it also throws that error. If this was previously working, then maybe something in the dependencies is causing problems

**Comment by alecxvs** [comment]
It broke as of 92e77617cf130283bd9f859b8ba0cfd1f7b7ccd2
But presumably master works, so I guess I'll diff with master and see what I changed that broke it

**Comment by alecxvs** [comment]
I think I found the cause. Webpack makes it hard to troubleshoot but it's basically complaining about this line:
`export class ZabbixDatasource extends DataSourceApi`
It's extending DataSourceApi but that class doesn't exist.

Since Grafana 6.5, DataSourceApi was moved to @grafana/data. But in my running instance, I'm running 6.4, so that class is actually found in @grafana/ui. Thus, it is undefined when webpack tries to require it.

So it seems that perhaps using the latest grafana packages and moving imports to @grafana/data is going to limit the minimum version of the plugin to Grafana 6.5.x?

**Comment by alecxvs** [comment]
I fixed the error by just going back to 6.4.5. I didn't realize that using different type versions could cause problems with different Grafana versions but it's pretty obvious now. 6.4.5 doesn't have the jest errors and the types are still in grafana/ui. There are some other miscellaneous frontend updates that I made in the process.

**Pull request closed without merging**